---
title: "Load and analyse HIM datasets"
author: "Mark Ziemann"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    fig_width: 7
    fig_height: 7
theme: cosmo
---

Source: TBA

## Introduction

```{r, packages}

baseDir=getwd()
dataDir=paste(baseDir,"/him_chart_idats/HIM_idats",sep="")

suppressPackageStartupMessages({
  library("missMethyl")
  library("limma")
  library("minfi")
  library("IlluminaHumanMethylation450kanno.ilmn12.hg19")
  library("IlluminaHumanMethylationEPICanno.ilm10b2.hg19")
  library("ruv")
  library("RColorBrewer")
  library("matrixStats")
  library("gplots")
  library("FlowSorted.Blood.450k")
  library("reshape2")
  library("ggplot2")
  library("DMRcate")
  library("FlowSorted.Blood.EPIC")
  library("mitch")
  library("kableExtra")
  library("vioplot")
})

#source("meth_functions.R")

```

## Load data

Load the annotation data and the Epic methylation data.

This analysis is to be conducted on Ubuntu with R4.

```{r,load_data1}

ann = getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b2.hg19)

ann_sub = ann[,c("chr","pos","strand","Name","Islands_Name",
    "Relation_to_Island","UCSC_RefGene_Name","UCSC_RefGene_Group")]

targets_gen = read.metharray.sheet(dataDir, pattern = "HIM_EPIC_sampleSheet.csv")
#targets$ID = paste(targets$Sample_Group,targets_gen$Sample_Name,sep=".")
rgSet = read.metharray.exp(targets = targets_gen)
sampleNames(rgSet) = targets_gen$Sample_Name

rgSet$Slide <- as.numeric(rgSet$Slide)
rgSet$Gender <- as.character(rgSet$Gender)
rgSet$Sex <- as.character(rgSet$Gender)
rgSet$Sample_Name<- as.character(rgSet$Sample_Name)

```

## Testing familial structure

```{r,snptest}

snpBetas = getSnpBeta(rgSet)
d = dist(t(snpBetas))
hr = hclust(d, method = "complete", members=NULL)
plot(hr)

```

## Quality control

```{r,qc1}

detP = detectionP(rgSet)
qcReport(rgSet, sampNames = targets_gen$Sample_Name,
  pdf="qc-report_him.pdf")
cols=brewer.pal(4,"Set1")

barplot(apply(detP,2,mean),
  col=as.numeric(factor(targets_gen$Sample_Name)),
  las=2,cex.names= 0.8, cex.axis=0.75,
  main="Mean detection p-values of probe signals",
  ylab="Mean detection p-value")

barplot(apply(detP,2,mean),
  col=as.numeric(factor(targets_gen$Sample_Name)),
  las=2,cex.names= 0.8, cex.axis=0.75,ylim=c(0,0.010),
  main="Mean detection p-values of probe signals",
  ylab="Mean detection p-value")

```

## Preprocessing

```{r,preprocess1}

mset.raw = preprocessRaw(rgSet)

```


## Data exploration

Using Multi-dimensional scaling (MDS) plots before filtering.

```{r,mds1}

mdsPlot(mset.raw, sampGroups = targets_gen$Sample_Name, 
  sampNames=targets_gen$Social_interaction_on_ADOS,legendPos="bottom")

mdsPlot(mset.raw, sampGroups = targets_gen$Sex, 
  sampNames=targets_gen$SampleID,legendPos="bottom")

```

## Cell type composition analysis

Try estimatecellcounts2.

```{r,celltypes2}

cells <- estimateCellCounts2(rgSet, referencePlatform= "IlluminaHumanMethylationEPIC", 
  returnAll = TRUE)

mset <- cells$normalizedData

cellCounts_new <- cells[[1]]
#plot cell type composition by sample group
a = cellCounts_new[targets_gen$Diagnosis == "0",]
b = cellCounts_new[targets_gen$Diagnosis == "1",]
c = cellCounts_new[targets_gen$Diagnosis == "2",]
age.pal <- brewer.pal(8,"Set1")

cellCounts_long <- melt(cellCounts_new, id = "celltype")

```

## Filtering

```{r,filt1}

detP <- detectionP(rgSet)

# exclude bad samples
sample_failed_probes <- apply(detP,2, function(x) { length(which(x>0.01)) / length(x) } )
barplot(sample_failed_probes[order(sample_failed_probes)])
bad_samples <- sample_failed_probes[sample_failed_probes>0.01]
bad_samples <- names(bad_samples)
good_samples <- setdiff(names(sample_failed_probes),bad_samples)
detP <- detP[,good_samples]
mset <- mset[,good_samples]
targets_gen <- targets_gen[which(targets_gen$Sample_Name %in% good_samples),]


# exclude bad probes
failed_prope_frac <- apply(detP,1, function(x) { length(which(x>0.01)) / length(x) } )
probes_to_keep <- names(which(failed_prope_frac<0.05))
detP <- detP[probes_to_keep,]
mset <- mset[which(rownames(mset) %in% probes_to_keep),]

dim(mset)
gmset <- mapToGenome(mset)

#remove SNPs
gmset_flt = dropLociWithSnps(gmset, snps = c("CpG", "SBE"))

#Removing cross-reactive probes
XURL="https://raw.githubusercontent.com/sirselim/illumina450k_filtering/master/EPIC/13059_2016_1066_MOESM1_ESM.csv"
Xreact <- read.csv(XURL)

#Xreact = read.csv(file="/group/canc2/puumba/Data/InfiniumData/NamithaData/Rprojects/Autism/Analysis_Sept11/EPIC_850k_crossreactiveProbes.csv", stringsAsFactors=FALSE)
#Xreact = read.csv(file="~/48639-non-specific-probes-Illumina450k.csv", stringsAsFactors=FALSE)
noXreact <-  !(featureNames(gmset) %in% Xreact$X)

gmset <- gmset[noXreact,]

#Removing probes on X and Y chromosomes
autosomes <- !(featureNames(gmset) %in% ann$Name[ann$chr %in% c("chrX","chrY")])
gmset_flt <- gmset[autosomes,]

#Relative log expression (RLE plot)
mvals = getM(gmset_flt)
medSq = apply(mvals, 1, median)
YSq = mvals - medSq

boxplot(YSq,outline=FALSE,ylim=c(-1.5,1.5), 
  ylab="Relative Log Methylation Value", 
  cols=as.character(factor(targets_gen$Social_interaction_on_ADOS,)) )

```

## MDS plots generation after filtering

```{r,mds2}

pal = brewer.pal(8, "Dark2")
mds1Sq = plotMDS(mvals, top=1000, gene.selection="common",dim.plot=c(1,2))
mds2Sq = plotMDS(mvals, top=1000, gene.selection="common",dim.plot=c(1,3))
mds3Sq = plotMDS(mvals, top=1000, gene.selection="common",dim.plot=c(2,3))
mds4Sq = plotMDS(mvals, top=1000, gene.selection="common",dim.plot=c(3,4))

plotMDS(mds1Sq, xlab="Dimension 1", ylab="Dimension 2",
  col=pal[as.factor(targets_gen$Diagnosis)],
  dim=c(1,2), labels=targets_gen$Sample_Name)
legend("bottomright",bg="white",col=pal,cex=.7,pch=1,legend=0:1)

plotMDS(mds2Sq, xlab="Dimension 1", ylab="Dimension 3",
  col=pal[as.factor(targets_gen$Diagnosis)],dim=c(1,3),
  labels=targets_gen$Sample_Name)
legend("bottomright",bg="white",col=pal,cex=.7,pch=1,legend=0:1)

plotMDS(mds3Sq, xlab="Dimension 2", ylab="Dimension 3",
  col=pal[as.factor(targets_gen$Diagnosis)],dim=c(2,3),
  labels=targets_gen$Sample_Name)
legend("bottomright",bg="white",col=pal,cex=.7,pch=1,legend=0:1)

plotMDS(mds4Sq, xlab="Dimension 3", ylab="Dimension 4",
  col=pal[as.factor(targets_gen$Diagnosis)],dim=c(3,4),
  labels=targets_gen$Sample_Name)
legend("bottomright",bg="white",col=pal,cex=.7,pch=1,legend=0:1)

```

## Principal Component Analysis (PCA)

```{r,pca1}

fit <- prcomp(t(mvals),center = TRUE, scale = TRUE,retx=TRUE)
loadings = fit$x
plot(fit,type="lines")
nGenes = nrow(mvals)
nSamples = ncol(mvals)

```

## Limma analysis looking for slide effects

Make a random variable for constrast.

```{r,reg1}

set.seed(42) ; rand <- sample(targets_gen$Sample_Name,45) 
targets_gen$rand <- targets_gen$Sample_Name %in% rand
rand <- targets_gen$rand
design <- model.matrix(~ rand )
fit <- lmFit(mvals, design)
fit2 <- eBayes(fit)
summary(decideTests(fit2))
top <- topTable(fit2,coef=ncol(design),num=Inf, sort.by = "P")


nsig <- sum(top$adj.P.Val < 0.05)
sum(top$P.Value< 0.05)
output <-merge(ann_sub,top,by.x="Name",by.y="row.names")
output <- output[order(output$P.Value),]
write.csv(output, file="limma_HIM_rand.csv",row.names=FALSE)
output <- subset(output,P.Value<1e-4)
head(output,30) %>% kbl() %>% kable_paper("hover", full_width = F)
saveRDS(design, "HIM_rand_des.rds")
saveRDS(mvals, "HIM_rand_mvals.rds")

```

## Session information

```{r,session}

sessionInfo()

```

